### PR TITLE
Devenv tweaks

### PR DESCRIPTION
**Pull request by ghaskins**
Development environment tweaks

Passes unit-tests.  Behave tests fail with 

```
Failing scenarios:
  peer_basic.feature:490  chaincode example02 with 4 peers and 1 membersrvc, issue #680 (State transfer) -- @1.1 Consensus Options
  peer_basic.feature:491  chaincode example02 with 4 peers and 1 membersrvc, issue #680 (State transfer) -- @1.2 Consensus Options
  peer_basic.feature:492  chaincode example02 with 4 peers and 1 membersrvc, issue #680 (State transfer) -- @1.3 Consensus Options
  peer_basic.feature:681  4 peers and 1 membersrvc, consensus fails if 2 backup replicas fail -- @1.1 Consensus Options

0 features passed, 1 failed, 0 skipped
19 scenarios passed, 4 failed, 0 skipped
484 steps passed, 4 failed, 27 skipped, 0 undefined
```

However, I do not believe these failures have anything to do with the changes as #680 has had intermittent failures for some time.  If this is a bad assumption, let me know.

Signed-off-by: Greg Haskins gregory.haskins@gmail.com

### CODE REVIEW

**Comment by ghaskins** [comment]
@rameshthoomu @bcbrock FYI, these changes form the basis of how we might begin to unify the various environments (vagrant, travisci, custom, etc).  It would be nice if we could find a way to define the baseimage version in one place (right now its primarily in devenv/Vagrantfile, but this is TBD.

File: devenv/Vagrantfile [hunk 1]
```diff
@@ -20,13 +20,6 @@ if not ENV.has_key?('GOPATH')
   abort("GOPATH is not set")
```
Comment (ghaskins) [inline devenv/Vagrantfile]: I suppose we don't need this check any more either.

**Comment by jyellick** [comment]
@ghaskins As of PR #997 which included some env var fixes, The #680 behave test was passing reliably for me.  If this is not the case, for you, could you open an issue and attach the logs from preferably the 'classic' pbft test?

**Comment by binhn** [comment]
@ghaskins could you also change the location `+NAME=openblockchain/baseimage`  to  `+NAME=hyperledger/baseimage`

**Comment by ghaskins** [comment]
@jyellick sure thing

@binhn absolutely.  Note we currently have the following hosted repositories in play
- https://atlas.hashicorp.com/obc - Vagrant images
- https://hub.docker.com/r/openblockchain/ - Docker images
- https://launchpad.net/~openblockchain - Personal Package Archive (PPA) - Not currently in use

All of those services are candidates for rebranding the organization.  Would you like me to rename/acquire all accordingly?

**Comment by ghaskins** [comment]
On further investigation, it seems someone has already created the hyperledger account on dockerhub 

https://hub.docker.com/u/hyperledger/

Therefore, I will not be able to post images there unless I am added to the ACL.  I will follow up with the others

**Comment by bcbrock** [comment]
@ghaskins I like the fact that you pulled out docker.sh into a separate script. If I understand correctly we should be able to make this script portable between architectures by having it choose the correct base image (once we define them).

**Comment by ghaskins** [comment]
@bcbrock yeah, i think there will be an opportunity to refine this concept further: $arch as you mentioned, but also to try to capture common things between vagrant/travis/custom/other environments, etc, too.

**Comment by binhn** [comment]
@ghaskins I'll ask on the TSC call today to see who owes that

**Comment by rameshthoomu** [comment]
@ghaskins I ran unit and behave tests and observed unit tests are passing and below behave scenarios are failing. 

```
Failing scenarios:
 peer_basic.feature:409  chaincode example02 with 4 peers and 1 membersrvc, issue #567 -- @1.4 Consensus Options
 peer_basic.feature:490  chaincode example02 with 4 peers and 1 membersrvc, issue #680 (State transfer) -- @1.1 Consensus Options
 peer_basic.feature:491  chaincode example02 with 4 peers and 1 membersrvc, issue #680 (State transfer) -- @1.2 Consensus Options
 peer_basic.feature:492  chaincode example02 with 4 peers and 1 membersrvc, issue #680 (State transfer) -- @1.3 Consensus Options

0 features passed, 1 failed, 0 skipped
19 scenarios passed, 4 failed, 0 skipped
499 steps passed, 4 failed, 12 skipped, 0 undefined
Took 16m35.361s

```

I could see scenarios #680 #567 are failing with latest fabric code..

Another Observation: While building an docker image using docker file placed in consensus/docker-compose-files i am getting below issues

**Inside Vagrant:**
vagrant@hyperledger-devenv:v0.0.9-7898461:/opt/gopath/src/github.com/hyperledger/fabric/consensus/docker-compose-files$ cat Dockerfile

```
from openblockchain/baseimage:latest
# Copy GOPATH src and install Peer
RUN mkdir -p /var/hyperledger/db
RUN mkdir -p /var/hyperledger/production
WORKDIR $GOPATH/src/github.com/hyperledger/fabric/
COPY . .
RUN CGO_CFLAGS=" " CGO_LDFLAGS="-lrocksdb -lstdc++ -lm -lz -lbz2 -lsnappy" go install && cp $GOPATH/src/github.com/hyperledger/fabric/core.yaml $GOPATH/bin
RUN cp $GOPATH/src/github.com/hyperledger/fabric/consensus/obcpbft/config.yaml $GOPATH/bin
# RUN CGO_CFLAGS=" " CGO_LDFLAGS="-lrocksdb -lstdc++ -lm -lz -lbz2 -lsnappy" go install
# RUN cd membersrvc && go install
```

vagrant@hyperledger-devenv:v0.0.9-7898461:/opt/gopath/src/github.com/hyperledger/fabric/consensus/docker-compose-files$ docker build -t <ImageName> .

```
Step 6 : RUN CGO_CFLAGS=" " CGO_LDFLAGS="-lrocksdb -lstdc++ -lm -lz -lbz2 -lsnappy" go install && cp $GOPATH/src/github.com/hyperledger/fabric/core.yaml $GOPATH/bin
 ---> Running in 0a0b02ccabbd
can't load package: package github.com/hyperledger/fabric: no buildable Go source files in /opt/gopath/src/github.com/hyperledger/fabric
The command '/bin/sh -c CGO_CFLAGS=" " CGO_LDFLAGS="-lrocksdb -lstdc++ -lm -lz -lbz2 -lsnappy" go install && cp $GOPATH/src/github.com/hyperledger/fabric/core.yaml $GOPATH/bin' returned a non-zero code: 1
```

**Outside vagrant (From Softlayer machine):**

Same Docker file as mentioned above

```
root@qa-peer02:/usr/lib/cgi-bin/github.com/hyperledger/fabric/consensus/docker-compose-files# 

"docker build -t <Image Name> ."

Sending build context to Docker daemon 19.46 kB
Step 1 : FROM openblockchain/baseimage:latest
Pulling repository docker.io/openblockchain/baseimage
Tag latest not found in repository docker.io/openblockchain/baseimage
```

if i modify tag to *_0.0.9 *_in docker file getting below error message:

```
Step 6 : RUN CGO_CFLAGS=" " CGO_LDFLAGS="-lrocksdb -lstdc++ -lm -lz -lbz2 -lsnappy" go install && cp $GOPATH/src/github.com/hyperledger/fabric/core.yaml $GOPATH/bin
 ---> Running in bb66015b7f24
/bin/sh: 1: go: not found
```

**Comment by ghaskins** [comment]
@jyellick it seems 567 and 680 are flaky in mainline as well.  @rameshthoomu reports that sometimes #724 is flaky too, but only within this PR so that needs more investigation.

@rameshthoomu as we discussed in DM, the Dockerfile problems I believe are a function of the "COPY , ," directive coupled with the recent change to move the Dockerfile down in the tree.  IOW, I don't think this is related to this patch.  And regarding running with explicit :0.0.9 tag, that is FAD as the image was not designed to be run directly.

**Comment by tuand27613** [comment]
@rameshthoomu   which commit are you on ?  @jyellick  committed c22268b and f561bae  in response to various folks seeing behave errors.
I checked in PR #1004 yesterday and behave tests were "behaving" :)

**Comment by ghaskins** [comment]
@tuand27613 its possible my branch that the PR is based on is stale (but conflict free), w.r.t. the fixes.  It is currently based on 

```
commit 89fa19065fa6a4cb053c61e1b82b25488f736898
Author: nickgaski <ngaski@us.ibm.com>
Date:   Tue Apr 5 12:35:23 2016 -0400

    Update README.md
```

Were the fixes put in after that?

**Comment by ghaskins** [comment]
@tuand27613 there are quite a few patches that went in between 89fa190 and origin/master:HEAD, so that is probably the explaination.

**Comment by srderson** [comment]
@ghaskins can we remove the Go 1.6 prerequisite and 'Set your GOPATH' step from https://github.com/hyperledger/fabric/blob/master/docs/dev-setup/devenv.md as part of this PR? Are they still required given this change?

**Comment by ghaskins** [comment]
@srderson I was debating whether the documentation should be updated:  I ultimately decided to let it be as anyone that wants to edit the code would need to have the GOPATH, etc, set up on the host to do so.  I figured if we had one set of instructions that said "do it vanilla" and another that said "oh yeah, and if you want to edit, do this" might be confusing.  Rather, the ones that don't like to RTFM will just git-clone and try to run it (maybe) and then seek out the directions later when they are trying to get more serious about working with the code.  The ones that like to follow directions will end up with an optimal setup.

That was my rationale, anyway.  I don't have a strong opinion about it.  Let me know if you think I should clean it up.

**Comment by srderson** [comment]
LGTM. Confirmed that all tests are passing. Thanks!

@ghaskins and I spoke on Slack and decided that the Go and $GOPATH requirements should not be removed from the docs.

**Comment by rameshthoomu** [comment]
@ghaskins @binhn - I have registered "hyperledger" account in DockerHub upfront to push docker images from TRAVIS.

**Pull request merged**